### ISSUE TITLE
Create an internal lint for detecting "non-Unicode-aware" `BytePos` manipulations

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#128790]
Inspired by https://github.com/rust-lang/rust/issues/128717#issuecomment-2270315205. CC @jieyouxu.

---

Since we recover from lexically invalid tokens that are Unicode-confusable with tokens that are lexically valid (e.g., *U+037E Greek Question Mark* → *U+003B Semicolon*; *U+066B Arabic Decimal Separator* → *U+002C Comma*), (suggestion) diagnostic code down the line generally *ought not* make assumptions about the length and/or position in bytes a given `Span` "maps to".

In reality however, all too often (suggestion) diagnostic code doesn't follow this when performing low-level "span manipulations" defaulting to hard-coded lengths and/or positions. The compiler contains a bunch of snippets like `+ BytePos(1)` or `- BytePos(1)` where the code guesses that code point before/after corresponds to a certain token/lexeme like `,`, `;`, `)` (which is fair). However, such code doesn't account for the aforesaid recovery which may have mapped a UTF-8 code unit with byte length > 1 to an ASCII character of length 1 which can lead to ICEs (internal assertions triggering or indexing/slicing at non-char-boundaries).

---

So it might we worth linting on these error-prone `BytePos` manipulations.
I don't know how feasible it'd be to implement *well* (i.e., low false positive rate) or how the exact rules should look like.

This issue may also serve as a tracking issue for eliminating this 'pattern' from the code base.

---

Uplifted from https://github.com/rust-lang/rust/issues/128717#issuecomment-2273735687:

It's so easy to find these kinds of ICEs:

1. One simply needs to look for `/(+|-) BytePos\(\d+\)/` inside `compiler/`,
2. Figure out which ASCII character is meant
3. Go to one's favorite Unicode table website that can list Unicode-confusables
4. Pick a confusable whose `.len_utf8()` is >1 and :boom:

---

**Example ICE**: #128717.

**Example ICE**: I just this found a minute ago while reviewing an unrelated PR:

This code uses a *Medium Right Parenthesis Ornament* (U+2769) which is confusable with *Right Parenthesis*.

```rs
fn f() {}

fn main() {
    f(0,1❩;
}
```

Leads to:

```
thread 'rustc' panicked at compiler/rustc_span/src/lib.rs:2119:17:
assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32
```

Code in question:

https://github.com/rust-lang/rust/blob/c9687a95a602091777e28703aa5abf20f1ce1797/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L1144

**Comment by jieyouxu** [comment]
As for suggestions:

- Suggestions should try to derive spans from existing spans using codepoint-boundary-aware operations available on `Span`, like `Span::{to, between, until}` etc.
- If suggestions really need to manipulate by codepoint granularity, consider using codepoint-boundary-aware helpers on `SourceMap`: `SourceMap::{start_point, end_point, next_point, span_through_char, span_until_whitespace, span_until_non_whitespace, span_until_char}`.
- If a suggestion really, really cannot get away with constructing the desired span from previous methods, then the `BytePos` manipulations need to carefully consider if a parser recovery scheme can reach the suggestion, or in the case of string literals and macros, can the user provide input that would cause codepoint boundary issues.

**Comment by compiler-errors** [comment]
Alternatively, if suggestions are often finding that a span needs to be recreated to match some subset of the AST (e.g. a keyword's span like `async` needs to be tracked for suggestions downstream), consider modifying the AST directly to track that span. But please only do this if it's *generally* useful, and not just needed for a single suggestion, and also be sure to audit the code for existing suggestions to simplify.

**Comment by Noratrieb** [comment]
Instead of writing a lint it sounds better to redesign the API to make this hard, like making the BytePos field private.
